### PR TITLE
[4416][ADD] purchase_order_line_payment_forecast

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
[4416](https://www.quartile.co/web#id=4416&cids=3&model=project.task&view_type=form)

### CODE REVIEW

File: purchase_order_line_payment_forecast/models/purchase_order_line.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+from dateutil.relativedelta import relativedelta
+
+from odoo import api, fields, models
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = "purchase.order.line"
+
+    price_subtotal_billed = fields.Monetary(
+        compute="_compute_price_subtotal", store=True
```
Comment (yostashiro) [inline purchase_order_line_payment_forecast/models/purchase_order_line.py]: To make the scope clearer.
```suggestion
        compute="_compute_price_subtotal_billed", store=True
```

File: purchase_order_line_payment_forecast/models/purchase_order_line.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                    ol.price_unit * ol.qty_invoiced
+                    for ol in line.order_id.order_line
+                    if ol.id != line.id and not ol.is_deposit
+                )
+                line.price_subtotal_billed = line.price_unit
+                line.price_subtotal_unbilled = (
+                    sum_billed_amount_other_lines - line.price_unit
+                )
+
+                continue
```
Comment (yostashiro) [inline purchase_order_line_payment_forecast/models/purchase_order_line.py]: I think my initial advice wasn't valid.  This should probably do.  Please double check.

```suggestion
                line.price_subtotal_billed = line.price_unit
                line.price_subtotal_unbilled = sum(line.invoice_lines.mapped("balance")) * -1
                continue
```

**Review by yostashiro — CHANGES_REQUESTED** [review]
Partial review.

File: purchase_order_line_payment_forecast/models/purchase_order_line.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        for line in self:
+            if line.is_deposit:
+                line.price_subtotal_billed = sum(
+                    line.invoice_lines.filtered(
+                        lambda l: l.move_id.state != "cancel"
+                    ).mapped("balance")
+                )
+                line.price_subtotal_unbilled = line.price_unit - abs(
+                    line.price_subtotal_billed
+                )
```
Comment (yostashiro) [inline purchase_order_line_payment_forecast/models/purchase_order_line.py]: The logic doesn't look correct.  It should be like this if I'm not mistaken.  Please check and confirm.  Note that the sum of price_subtotal_unbilled is supposed to be the amount they should expect to pay after the current AP balance is cleared.

```suggestion
                line.price_subtotal_billed = 0.0
                line.price_subtotal_unbilled =  sum(
                    line.invoice_lines.filtered(
                        lambda l: l.move_id.state != "cancel"
                    ).mapped("balance")
                ) * -1
```

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: purchase_order_line_payment_forecast/models/purchase_order_line.py (hunk 3 above)
Comment (AungKoKoLin1997) [reply purchase_order_line_payment_forecast/models/purchase_order_line.py]: I checked and I think your suggestion is correct but I am wondering price_subtotal_billed should always be `0.00` for deposit line.

Comment (yostashiro) [reply purchase_order_line_payment_forecast/models/purchase_order_line.py]: @AungKoKoLin1997 How about this, then?
```.py
                line.price_subtotal_billed =  sum(
                    line.invoice_lines.filtered(
                        lambda l: l.move_id.state != "cancel"
                    ).mapped("balance")
                )
                line.price_subtotal_unbilled = line.price_subtotal_billed * -1
```

**Comment by kanda999** [comment]
@AungKoKoLin1997 
When I install this moudule, Expected Payment Date will not be updated.  
Why is it happened?

**Comment by AungKoKoLin1997** [comment]
@kanda999  Please review my last commit.

File: purchase_order_line_payment_forecast/hooks.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# Copyright 2024 Quartile (https://www.quartile.co)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo.tools.sql import column_exists
+
+
+def pre_init_hook(cr):
```
Comment (yostashiro) [inline purchase_order_line_payment_forecast/hooks.py:10]: @AungKoKoLin1997 Can you please comment the intention of this?

**Review by yostashiro — APPROVED** [review]
Code review.

@kanda999 Please double-check that expected payment date is correctly set upon installation.

**Review by kanda999 — APPROVED** [review]
Code review.
And expected payment date is correctly set during installation.

**Pull request merged**